### PR TITLE
pkg/ipam: added a max-allocate flag

### PR DESCRIPTION
**Pull request by mvisonneau**
The current behaviour of the IPAM is to always attempt to allocate IP addresses according to the `pre-allocate` flag. This change will allow endusers to override this behaviour if they want to by specifying a `max-allocate` flag.

This could prevent the **cilium-operator** from looping when there is no ENIs/IPs that can possibly allocated when a node is approaching its limits. 

Example using a t3a.medium with 15 pods running and pre-allocate set to 8 (default):

```
level=info msg="Synchronized ENI information" numENIs=6 numSecurityGroups=5 numSubnets=12 numVPCs=1 subsys=eni
level=info msg="Resolving IP deficit of node" available=18 instanceID=i-xxxxxxx name=ip-10-0-0-0.eu-west-1.compute.internal remainingInterfaces=0 subsys=eni toAlloc=5 used=15
level=info msg="Synchronized ENI information" numENIs=6 numSecurityGroups=5 numSubnets=12 numVPCs=1 subsys=eni
level=info msg="Resolving IP deficit of node" available=18 instanceID=i-xxxxxxx name=ip-10-0-0-0.eu-west-1.compute.internal remainingInterfaces=0 subsys=eni toAlloc=5 used=15
level=info msg="Synchronized ENI information" numENIs=6 numSecurityGroups=5 numSubnets=12 numVPCs=1 subsys=eni
level=info msg="Resolving IP deficit of node" available=18 instanceID=i-xxxxxxx name=ip-10-0-0-0.eu-west-1.compute.internal remainingInterfaces=0 subsys=eni toAlloc=5 used=15
level=info msg="Synchronized ENI information" numENIs=6 numSecurityGroups=5 numSubnets=12 numVPCs=1 subsys=eni
level=info msg="Resolving IP deficit of node" available=18 instanceID=i-xxxxxxx name=ip-10-0-0-0.eu-west-1.compute.internal remainingInterfaces=0 subsys=eni toAlloc=5 used=15
```

```release-note
added a `max-allocate` flag on **pkg/ipam** to control the maximum amount of IPs being allocated to a node
```

### CODE REVIEW

**Comment by mvisonneau** [comment]
as an example, with only one instance reaching the limits, we can get up to ~800k loops executed over 24 hours

image

I believe it could also be interesting to get some kind of backoff mechanism in place in order to deal with these situations

**Comment by aanm** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Review by aanm — APPROVED** [review]
changes LGTM @tgraf do you want to take a look as well?

File: pkg/ipam/node.go
Comment (qmonnet) [inline pkg/ipam/node.go]: Nit: Can we remove this commented, but otherwise empty line?

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks good to me, with a couple of minor nitpicks.

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (qmonnet) [inline pkg/k8s/apis/cilium.io/v2/register.go]: Nit: Some lines are already longer than this in the file, so it's maybe not worth wrapping for just one word? (Here and on the other occurrence below)

[Sorry, somehow I failed to add this comment to the previous review.]

**Review by qmonnet — CHANGES_REQUESTED** [review]

File: pkg/ipam/node.go
Comment (mvisonneau) [reply pkg/ipam/node.go]: sounds good, done! 👍

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (mvisonneau) [reply pkg/k8s/apis/cilium.io/v2/register.go]: indeed 🤦‍♂ , 👍

**Review by qmonnet — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (errordeveloper) [inline pkg/k8s/apis/cilium.io/v2/register.go:420]: Once you are touching these, would you mind adding a description to `pre-allocate` please?

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (errordeveloper) [inline pkg/k8s/apis/cilium.io/v2/register.go:479]: ...and here.

**Review by errordeveloper — COMMENTED** [review]
Thanks for this PR, I think it's very much ready to go, if you wouldn't mind adding descriptions to `pre-allocate` while rebasing, it would be really nice!

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (mvisonneau) [reply pkg/k8s/apis/cilium.io/v2/register.go:420]: sure! I have also done it for `max-above-watermark` as well 👌

File: pkg/k8s/apis/cilium.io/v2/register.go
Comment (mvisonneau) [reply pkg/k8s/apis/cilium.io/v2/register.go:479]: 👍

**Review by errordeveloper — APPROVED** [review]
Thanks for improving the docs @mvisonneau!

**Comment by errordeveloper** [comment]
test-me-please

**Comment by errordeveloper** [comment]
will wait for @qmonnet's review!

**Review by qmonnet — APPROVED** [review]
Sorry for the delay, looks good!

**Comment by mvisonneau** [comment]
I could not find what was wrong with the checks 🤔

Do you want me to rebase as well btw?

**Comment by joestringer** [comment]
The checks have expired, I'll kick them again.

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
Rebasing is typically only necessary when Github at the end of the page says there are merge conflicts. In this case it's not necessary, when we re-run CI, the CI will rebase your PR when it runs the tests.

**Comment by mvisonneau** [comment]
I am not sure to understand why the `Cilium-Ginkgo-Tests` are failing 🤔

**Comment by joestringer** [comment]
Looking at https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/18572/, I see that exactly one test failed. The `Cilium-Ginkgo-Tests` runs two parallel test runs against different k8s versions, and `Cilium-Tests-With-Kernel` does the same for another version + kernel. Since only one test failed across the three runs, it's not a consistent failure. That suggests a flake.

If it's a flaky CI test, then the question becomes: How likely is it that it exists already and how likely is it that the flake is introduced by this PR? One way we can track this is by searching for (or filing) issues using the "Failing Test" template (seen when you click the "New issue" button). We typically name testing issues with `CI: Nameoftestthatfailed", potentially with a symptom as a suffix, eg in this case `CI: K8sServicesTest Checks service across nodes Tests NodePort BPF Tests with vxlan Tests NodePort with externalTrafficPolicy=Local: command terminated with exit code 28`. The template states what is expected there, usually copying the contents of the failure from the failure page, and attaching the sysdump zip from the CI. This helps contributors like you by providing proof that others have seen the CI failure before, and helps the maintainers track which flakes are causing disruption to development (and how frequently). Deeper debugging can then be performed by pulling the sysdump zip and digging through logs and other state.

AFAICT from the changes in this PR, it seems unlikely that this PR introduced the CI issue since it's making IPAM changes and only a nodeport services test failed.

**Comment by mvisonneau** [comment]
thanks a lot for this very insightful explanation @joestringer ! 🙇

**Comment by aanm** [comment]
test-me-please

**Pull request merged**
